### PR TITLE
Access aggregate's child applications' request mapping/endpoint web endpoints

### LINKED ISSUE #702: Applications that are part of an aggregate should be able to expose web endpoints
If the child applications of the aggregate application require web endpoints to be enabled (ex: aggregate source application being http source), those endpoints need to be enabled.

Here is the SO issue for more info: http://stackoverflow.com/questions/40295959/springcloudstream-httpsource-is-returning-404

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Use namespace to prefix the endpoints' paths
 - Lazy initialize EndpointHandlerMapping once child application context is updated with its actuator endpoints
 - Disable actuator endpoints for aggregator parent context

Resolves #702

### CODE REVIEW

**Comment by ilayaperumalg** [comment]
Please note that this PR exposes the spring boot actuator endpoints of the child applications' and not the MVC request mappings of the child applications.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/AggregateApplicationBuilder.java [hunk 1]
```diff
@@ -386,15 +424,7 @@ void embed() {
 							e);
 				}
 			}
-			// Register metrics if JMX enabled and exporter avalable
-			if (BeanFactoryUtils.beansOfTypeIncludingAncestors(AggregateApplicationBuilder.this.parentContext,
```
Comment (ilayaperumalg) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/AggregateApplicationBuilder.java]: Given all the metrics are now exposed via `/<namespace>/metrics` endpoint, I believe this explicit registration of metrics can be removed.

**Comment by ilayaperumalg** [comment]
The latest update to this PR takes care of mapping the `RequestMapping` from the child context to the aggregate parent.

**Comment by ilayaperumalg** [comment]
Rebased and updated the PR with the changes needed to not enforcing the `EnableAutoConfiguration` on Aggregate ParentConfiguration.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/AggregateApplicationUtils.java [hunk 2]
```diff
@@ -59,10 +65,16 @@ static String getDefaultNamespace(String appClassName, int index) {
...
+		// Child context needs to enable web MVC configuration and web enabled to obtain the MVC request mapping in the child applications
+		return new SpringApplicationBuilder(new Object[]{app, EmbeddedServletContainerAutoConfiguration.class,
+				EndpointWebMvcManagementContextConfiguration.class, WebMvcAutoConfiguration.EnableWebMvcConfiguration.class,
+				ServerPropertiesAutoConfiguration.class})
+				.web(true).main(app)
+				.bannerMode(Mode.OFF)
 				.properties("spring.jmx.default-domain=" + namespace)
 				.properties(
 						InternalPropertyNames.NAMESPACE_PROPERTY_NAME + "=" + namespace)
+				.properties("server.port=-1")
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/AggregateApplicationUtils.java]: Still not OK with doing this. If starting one web server per application is the only way we can get this done, then we should drop the idea altogether and not support embedding of web apps.

Comment (ilayaperumalg) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/aggregate/AggregateApplicationUtils.java]: In order to expose the web endpoints, the child applications need to be web `enabled` but here we make sure they don't expose the `server.port`. One possible way is to introduce a property to let the user decide whether to expose the web endpoints for the child application or not. That way, we will start web application context only for those child apps that have their web endpoints (request mappings + actuator endpoints) mapped to parent aggregate application.

**Comment by mbogoevici** [comment]
Also it would be good to have some documentation on how this process works.

**Comment by ilayaperumalg** [comment]
Rebased.

**Comment by mbogoevici** [comment]
I've been reviewing this and the approach per se looks OK, the only caveat is that it couples the core even tighter to a dependency on the `web` module of Spring Boot. I think we can make all this stuff optional on whether web classes are on the classpath (otherwise it is not applicable anyway).

**Comment by ilayaperumalg** [comment]
@mbogoevici Given we already have `spring-boot-starter-web` in core [classpath](https://github.com/spring-cloud/spring-cloud-stream/blob/master/spring-cloud-stream/pom.xml#L21), the web classes are already there. 

Also, We [enable](https://github.com/spring-cloud/spring-cloud-stream/pull/776/files#diff-e6a31d66d78de3178c1715226b5a4117R233) the `web` configuration only when the web is enabled for the aggregate application. The child application contexts' web is disabled as well.

**Comment by mbogoevici** [comment]
@ilayaperumalg 
>Given we already have spring-boot-starter-web in core classpath, the web classes are already there.

That is a contentious point already and we might want to remove that dependency in the future. It is not effectively required at the moment. Currently it is possible to exclude it and still get a functional application (albeit without web actuator endpoints). With this change, it will be impossible to remove `spring-boot-starter-web` for an aggregate application (`AggregateApplicationBuilder` will probably not load).

**Comment by ilayaperumalg** [comment]
@mbogoevici I agree, Didn't realize the strategy to remove `web` starter dependency. I will update this PR.

**Comment by ilayaperumalg** [comment]
Given the complexity involved in maintaining the `Endpoint` handler mappings and the future efforts to maintain the compatibility with Spring Boot changes, we are closing this PR.

The work around to fix this issue would be to run the `source` application as a bean outside the aggregate application and have `outbound` channel of the `source` manually connect to the `inbound` channel of the aggregate application. This way, the `source` application acts as a `gateway` to receive the incoming request endpoints while the `binding` between the `source` and the aggregate application is wired inside the same application.

**Pull request closed without merging**